### ISSUE TITLE
Question about total_ppo_epoch

### ISSUE THREAD

**Issue by BaleChen** [huggingface/trl#712]
Hi! I encountered a question about this [line](https://github.com/huggingface/trl/blob/bcf45a2d7acaa43a83c98dc7eb0f38febc1a4058/trl/trainer/ppo_config.py#L208C75-L208C75
) of code in ppo trainer:
`self.total_ppo_epochs = int(np.ceil(self.steps / self.batch_size))`

Can someone explain to me what does steps / batch_size mean and how is it the total_ppo_epoch? The rationale behind this line is not very clear to me. Thanks!

**Comment by malik-ali** [comment]
As far as I understand it, it's just a constant for you to use in your training loop so that the total number of gradient updates of your model is equal to `steps`.

So if you have
```
for epoch in range(ppo_trainer.total_ppo_epochs):
   for batch in ppo_trainer.dataloader:
      # compute loss and do optimizer step
```

then this code will run `steps` many optimizer steps.

**Comment by BaleChen** [comment]
If I understand correctly, to make the total number of update steps equal to `steps`, total_ppo_epochs should be `int(np.ceil(self.steps / len(self.dataloader)))` instead of dividing by the batch_size. In your code snippet, there are `total_ppo_epoch * len(ppo_trainer.dataloader)` update steps. Am I right?

**Comment by BaleChen** [comment]
Also, I was looking at stack_llama/scripts/rl_training.py. They wrote [here](https://github.com/huggingface/trl/blob/25d6700c5e6db96a722942c9c91594240792346d/examples/research_projects/stack_llama/scripts/rl_training.py#L241C1-L243C14).

```python
for epoch, batch in tqdm(enumerate(ppo_trainer.dataloader)):
    if epoch >= config.total_ppo_epochs:
        break
``` 

It only runs through the dataset once and calls each PPO step on a batch an `epoch`. I'm a bit confused here. Could someone also clarify this?

**Comment by BaleChen** [comment]
No worries, it's indeed confusing. 

@younesbelkada @lvwerra Hi both! Any good advice? Thank you for your time.

**Comment by BaleChen** [comment]
Sounds good. Just fyi, I currently use in the outer loop `for epoch in range(n_epochs):` where `n_epochs = args.steps / len(ppo_trainer.dataloader)` as a workaround.
